packages/dual-channel/src/build-code/full-width-wrapper.js

```diff
@@ -0,0 +1,102 @@
+import PropTypes from 'prop-types'
+import React, { useState, useEffect, useCallback } from 'react'
+// lodash
+import debounce from 'lodash/debounce'
+
+const _ = {
+  debounce,
+}
+
+function getViewportWidth() {
+  if (document && document.documentElement) {
+    return document.documentElement.clientWidth || 0
+  }
+  return 0
+}
+
+/**
+ * get the distance from the element to viewport
+ *
+ * @param {HTMLElement} element
+ * @returns {number}
+ */
+function getXRelatedToViewport(element) {
+  if (element) {
+    const parent = element.parentElement
+    if (!parent) {
+      return element.getBoundingClientRect().left || 0
+    }
+    const parentComputedStyle = window.getComputedStyle(parent)
+    const parentPaddingLeft =
+      parseFloat(parentComputedStyle.getPropertyValue('padding-left')) || 0
+    const parentBorderLeft =
+      parseFloat(parentComputedStyle.getPropertyValue('border-left')) || 0
+    return (
+      parent.getBoundingClientRect().left +
+        parentPaddingLeft +
+        parentBorderLeft || 0
+    )
+  }
+  return 0
+}
+
+const defaultWidth = '95vw'
+const defaultXRelatedToViewport = 0
+
+export default function FullWidthWrapper(props) {
+  const { full, children } = props
+
+  if (!full) {
+    return children
+  }
+
+  const [xRelatedToViewport, setXRelatedToViewport] = useState(0)
+  const [viewportWidth, setViewportWidth] = useState(getViewportWidth)
```

Comment:
Since `useState` receives the initial value, is it `useState(getViewportWidth())`  ?

Reply:
It's a lazy state initiate function that only invoked at first render
Ref: https://zh-hant.reactjs.org/docs/hooks-reference.html#lazy-initial-state

Reply:
but it's okay to use `useState(getViewportWidth())` also, since `getViewportWidth` does not have heavy calculation